File: src/domain/projects.ts
import * as fs from "node:fs";
import { isAbsolute } from "node:path";
import { newId } from "../ids.js";
import type { EventBody, LogEvent } from "../storage/events.js";
import type { StateIndex } from "../storage/index.js";
import { ToolError } from "./errors.js";

export interface ProjectRef {
  project_id: string;
  root: string;
}

/**
 * Resolves a caller-supplied path to a canonical project record, creating the
 * project on first sight. Projects are keyed by normalized (real) root path;
 * other observed paths become aliases.
 */
export function resolveProject(
  index: StateIndex,
  record: (body: EventBody) => LogEvent,
  projectPath: string,
): ProjectRef {
  if (!isAbsolute(projectPath)) {
    throw new ToolError("invalid_request", `project path must be absolute: ${projectPath}`);
  }
  let root: string;
  try {
    root = fs.realpathSync(projectPath);
  } catch {
    throw new ToolError("invalid_request", `project path does not exist: ${projectPath}`);
  }
  if (!fs.statSync(root).isDirectory()) {
    throw new ToolError("invalid_request", `project path is not a directory: ${projectPath}`);
  }

  const byAlias = index.db
    .prepare(
      `SELECT p.id, p.root FROM project_aliases a JOIN projects p ON p.id = a.project_id
       WHERE a.path = ?`,
    )
    .get(root) as { id: string; root: string } | undefined;
  if (byAlias) {
    if (projectPath !== root) recordAlias(index, record, byAlias.id, projectPath);
    return { project_id: byAlias.id, root: byAlias.root };
  }

  const projectId = newId("proj");
  record({ type: "project.created", project_id: projectId, root });
  if (projectPath !== root) recordAlias(index, record, projectId, projectPath);
  return { project_id: projectId, root };
}

function recordAlias(
  index: StateIndex,
  record: (body: EventBody) => LogEvent,
  projectId: string,
  path: string,
): void {
  const existing = index.db
    .prepare("SELECT project_id FROM project_aliases WHERE path = ?")
    .get(path) as { project_id: string } | undefined;
  if (!existing) record({ type: "project.alias-added", project_id: projectId, path });
}

File: rust/src/domain/projects.rs
//! Resolving a caller-supplied path to a project record.

use std::fs;
use std::path::Path;

use rusqlite::OptionalExtension;

use super::errors::ToolError;
use crate::ids::{IdPrefix, new_id};
use crate::storage::events::EventBody;
use crate::storage::store::Store;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ProjectRef {
    pub project_id: String,
    pub root: String,
}

/// Resolves a caller-supplied path to a canonical project record, creating the
/// project on first sight. Projects are keyed by normalized (real) root path;
/// other observed paths become aliases. Takes the locked store so the lookup
/// and the create happen under one lock: two tasks resolving a new project at
/// once must not both create it.
pub fn resolve_project(store: &mut Store, project_path: &str) -> Result<ProjectRef, ToolError> {
    if !Path::new(project_path).is_absolute() {
        return Err(ToolError::invalid_request(format!(
            "project path must be absolute: {project_path}"
        )));
    }
    let root = fs::canonicalize(project_path).map_err(|_| {
        ToolError::invalid_request(format!("project path does not exist: {project_path}"))
    })?;
    if !root.is_dir() {
        return Err(ToolError::invalid_request(format!(
            "project path is not a directory: {project_path}"
        )));
    }
    let root = root.to_string_lossy().into_owned();

    let by_alias: Option<(String, String)> = store
        .index
        .db
        .query_row(
            "SELECT p.id, p.root FROM project_aliases a JOIN projects p ON p.id = a.project_id
             WHERE a.path = ?",
            [&root],
            |row| Ok((row.get(0)?, row.get(1)?)),
        )
        .optional()?;
    if let Some((project_id, root_of_record)) = by_alias {
        if project_path != root {
            record_alias(store, &project_id, project_path)?;
        }
        return Ok(ProjectRef { project_id, root: root_of_record });
    }

    let project_id = new_id(IdPrefix::Project);
    store
        .record(EventBody::ProjectCreated { project_id: project_id.clone(), root: root.clone() })?;
    if project_path != root {
        record_alias(store, &project_id, project_path)?;
    }
    Ok(ProjectRef { project_id, root })
}

fn record_alias(store: &mut Store, project_id: &str, path: &str) -> Result<(), ToolError> {
    let existing: Option<String> = store
        .index
        .db
        .query_row("SELECT project_id FROM project_aliases WHERE path = ?", [path], |row| {
            row.get(0)
        })
        .optional()?;
    if existing.is_none() {
        store.record(EventBody::ProjectAliasAdded {
            project_id: project_id.to_string(),
            path: path.to_string(),
        })?;
    }
    Ok(())
}
